File: ios/Plugin/KlarnaKcoPlugin.swift
import Foundation
import Capacitor
import KlarnaMobileSDK

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(KlarnaKcoPlugin)
public class KlarnaKcoPlugin: CAPPlugin {
    private var implementation: KlarnaKco?
    private var config = KlarnaKcoConfig()

    @objc
    func initialize(_ call: CAPPluginCall) {
        self.klarnaKcoConfig()
        self.implementation = KlarnaKco(plugin: self, config: self.config)
        call.resolve(["status": 200, "message": "ok"])
    }
    
    @objc
    func setSnippet(_ call: CAPPluginCall) {
        guard let implementation = self.implementation else {
            call.reject("Klarna plugin not initialized")
            return
        }

        let result = implementation.setSnippet(snippet: call.getString("snippet", ""))
        if result.status {
            call.resolve(["status": result.status, "message": result.message])
        } else {
            call.reject(result.message)
        }
    }
    
    @objc
    func open(_ call: CAPPluginCall) {
        guard let implementation = self.implementation else {
            call.reject("Klarna plugin not initialized")
            return
        }
        
        let result = implementation.open()
        if result.status {
            call.resolve(["status": result.status, "message": result.message])
            return
        } else {
            call.reject(result.message)
        }
    }
    
    @objc
    func close(_ call: CAPPluginCall) {
        guard let implementation = self.implementation else {
            call.reject("Klarna plugin not initialized")
            return
        }
        
        let result = implementation.close()
        if result.status {
            call.resolve(["status": result.status, "message": result.message])
        } else {
            call.reject(result.message)
        }
    }
    
    @objc
    func resume(_ call: CAPPluginCall) {
        guard let implementation = self.implementation else {
            call.reject("Klarna plugin not initialized")
            return
        }

        implementation.resume()
        call.resolve(["result": "invoked"])
    }
    
    @objc
    func suspend(_ call: CAPPluginCall) {
        guard let implementation = self.implementation else {
            call.reject("Klarna plugin not initialized")
            return
        }

        implementation.suspend()
        call.resolve(["result": "invoked"])
    }
    
    @objc
    func destroy(_ call: CAPPluginCall) {
        guard let implementation = self.implementation else {
            call.reject("Klarna plugin not initialized")
            return
        }

        implementation.destroy()
        call.resolve()
    }
}

extension KlarnaKcoPlugin {
    private func klarnaKcoConfig() {
        config.iosReturnUrl = URL(string: getConfig().getString("iosReturnUrl", "")!)
        config.handleValidationErrors = getConfig().getBoolean("handleValidationErrors", false)
        config.handleEPM = getConfig().getBoolean("handleEPM", false)
        
        switch getConfig().getString("region", "") {
        case "eu":
            self.config.region = KlarnaRegion.eu
            break
        case "na":
            config.region = KlarnaRegion.na
            break
        case "oc":
            config.region = KlarnaRegion.oc
            break
        default:
            break
        }
        
        switch getConfig().getString("environment", "") {
        case "demo":
            config.environment = KlarnaEnvironment.demo
            break
        case "playground":
            config.environment = KlarnaEnvironment.playground
            break
        case "staging":
            config.environment = KlarnaEnvironment.staging
            break
        case "production":
            config.environment = KlarnaEnvironment.production
            break
        default:
            break
        }
        
        switch getConfig().getString("loggingLevel", "") {
        case "verbose":
            config.loggingLevel = KlarnaLoggingLevel.verbose
            break
        case "error":
            config.loggingLevel = KlarnaLoggingLevel.error
            break
        case "off":
            config.loggingLevel = KlarnaLoggingLevel.off
            break
        default:
            break
        }
        
        switch getConfig().getString("theme", "") {
        case "light":
            config.theme = KlarnaTheme.light
            break
        case "dark":
            config.theme = KlarnaTheme.dark
            break
        case "automatic":
            config.theme = KlarnaTheme.automatic
            break
        default:
            break
        }
    }
}

File: android-oldish/src/main/java/com/capacitor/community/klarna/kco/KlarnaKcoPlugin.kt
package com.capacitor.community.klarna.kco

import com.getcapacitor.*
import com.getcapacitor.annotation.CapacitorPlugin
import com.klarna.mobile.sdk.api.KlarnaEnvironment
import com.klarna.mobile.sdk.api.KlarnaLoggingLevel
import com.klarna.mobile.sdk.api.KlarnaRegion

@CapacitorPlugin(name = "KlarnaKco")
class KlarnaKcoPlugin : Plugin() {
    private var implementation: KlarnaKco? = null
    private val config = KlarnaKcoConfig()

    @PluginMethod
    fun initialize(call: PluginCall) {
        setKlarnaKcoConfig()
        implementation = KlarnaKco(config, this)
        call.resolve()
    }

    @PluginMethod
    fun setSnippet(call: PluginCall) {
        val implementation = implementation
        if (implementation == null) {
            call.reject("Klarna plugin not initialized")
            return
        }
        try {
            val result = implementation.setSnippet(call.getString("snippet", ""))
            if (result != null) {
                val status = result.getBoolean("status", false)
                if (status != null && status) call.resolve(result) else call.reject(result.getString("message"))
            } else call.reject("Error in getting the result")
        } catch (e: Exception) {
            call.reject("An error occurred while setting Klarna checkout snippet: " + e.message)
            Logger.error(e.message, e)
        }
    }

    @PluginMethod
    fun open(call: PluginCall) {
        val implementation = implementation
        if (implementation == null) {
            call.reject("Klarna plugin not initialized")
            return
        }
        try {
            val result = implementation.open()
            if (result != null) {
                val status = result.getBoolean("status", false)
                if (status != null && status) call.resolve(result) else call.reject(result.getString("message"))
            } else call.reject("Error in getting the result")
        } catch (e: Exception) {
            call.reject("An error occurred while opening Klarna checkout view: " + e.message)
            Logger.error(e.message, e)
        }
    }

    @PluginMethod
    fun close(call: PluginCall) {
        val implementation = implementation
        if (implementation == null) {
            call.reject("Klarna plugin not initialized")
            return
        }
        try {
            val result = implementation.close()
            if (result != null) {
                val status = result.getBoolean("status", false)
                if (status != null && status) call.resolve(result) else call.reject(result.getString("message"))
            } else call.reject("Error in getting the result")
        } catch (e: Exception) {
            call.reject("An error occurred while closing Klarna checkout view: " + e.message)
            Logger.error(e.message, e)
        }
    }

    @PluginMethod
    fun resume(call: PluginCall) {
        val implementation = implementation
        if (implementation == null) {
            call.reject("Klarna plugin not initialized")
            return
        }
        implementation.resume()
        val ret = JSObject()
        ret.put("result", "invoked")
        call.resolve(ret)
    }

    @PluginMethod
    fun suspend(call: PluginCall) {
        val implementation = implementation
        if (implementation == null) {
            call.reject("Klarna plugin not initialized")
            return
        }
        implementation.suspend()
        val ret = JSObject()
        ret.put("result", "invoked")
        call.resolve(ret)
    }

    @PluginMethod
    fun destroy(call: PluginCall) {
        val implementation = implementation
        if (implementation == null) {
            call.reject("Klarna plugin not initialized")
            return
        }
        implementation.destroy()
        call.resolve()
    }

    fun notify(key: String, data: JSObject) {
        this.notifyListeners(key, data);
    }

    private fun setKlarnaKcoConfig() {
        val androidReturnUrl = getConfig().getString("androidReturnUrl", config.androidReturnUrl)
        config.androidReturnUrl = androidReturnUrl
        val handleValidationErrors = getConfig().getBoolean("handleValidationErrors", config.handleValidationErrors)
        config.handleValidationErrors = handleValidationErrors
        val handleEpm = getConfig().getBoolean("handleEPM", config.handleEPM)
        config.handleEPM = handleEpm
        val region = getConfig().getString("region", "")
        when (region) {
            "na" -> config.region = KlarnaRegion.NA
            "oc" -> config.region = KlarnaRegion.OC
            "eu" -> config.region = KlarnaRegion.EU
            else -> {}
        }
        val environment = getConfig().getString("environment", "")
        when (environment) {
            "demo" -> config.environment = KlarnaEnvironment.DEMO
            "playground" -> config.environment = KlarnaEnvironment.PLAYGROUND
            "staging" -> config.environment = KlarnaEnvironment.STAGING
            "production" -> config.environment = KlarnaEnvironment.PRODUCTION
            else -> {}
        }
        val loggingLevel = getConfig().getString("loggingLevel", "")
        when (loggingLevel) {
            "verbose" -> config.loggingLevel = KlarnaLoggingLevel.Verbose
            "error" -> config.loggingLevel = KlarnaLoggingLevel.Error
            "off" -> config.loggingLevel = KlarnaLoggingLevel.Off
            else -> {}
        }
    }
}